### ISSUE TITLE
tf.identity op on parameters in Distribution "breaks" GradientTape

### ISSUE THREAD

**Issue by mayou36** [tensorflow/probability#285]
In the `__init__` of `Distributions`, the parameters are stored with a `tf.identity` op (e.g. Normal: `self._loc = tf.identity(loc)`).

This does not allow to directly use GradientTape, as when using any method of `Distribution`, it only accesses the Tensor and not the Variable.

Example:
```python
import tensorflow_probability.python.distributions as tfd

var1, var2 = tf.Variable(1., use_resource=True), tf.Variable(2., use_resource=True)
normal = tfd.Normal(loc=var1, scale=var2)
with tf.GradientTape() as tape:
    prob = normal.prob(42.)
print("empty:", tape.watched_variables())  # nothing recorded

with tf.GradientTape() as tape2:
    prob2 = normal.parameters['loc'] * 42.  # this record the variable
print("not empty:", tape2.watched_variables())
```

The actual usecase is to have `custom_gradient`, that works internally with `tf.GradientTape`. In the current state, this can't be used together with the distributions.

One question: why, in the first place, is there even a `tf.identity` and not just the raw variable saved as `self._loc`?
And of course, can this somehow be changed? It's a deal-breaker for us, as we need a custom gradient on a combination of things. But one Distribution in the whole calculation breaks it.

**Comment by brianwa84** [comment]
Can you put the instantiation of the Normal inside the tape?

**Comment by mayou36** [comment]
No. I use a while-loop (#181 ) to generate a value and it's derivative on the fly. Therefore, all I get is a function, evaluate it on random samples.

The ironie is: it would work with `tf.gradients(dist.prob(x), variables)` _in principle_ inside the grad_fn, but `custom_gradient` does not pass through the `variables` with respect to whom the derivative should be taken through but fidles around itself (by using the `tf.GradientTape` to tape the Variables during the function execution).

A change on `tf.GradientTape` seems unlikely to me, however I also don't quite see the purpose of the tf.identity. As I build pdfs myself, _not_ using tf.identity, I am quite interested to know why this is there? What's the purpose of that?

**Comment by brianwa84** [comment]
The identity should probably be tf.convert_to_tensor, for consistency with be other code, but I think that would still be a problem for you. It's possible we could special case Variable parameters.

I'm still a bit unclear on how you are trying to use custom_gradient and why the while loop is relevant. Can you get away with using tensors instead of Variables, or do you need something that can checkpoint/persist across session.run calls?

**Comment by mayou36** [comment]
I need something persistent across session.run calls, yes.

Hm, a blind convert_to_tensor? Why doesn't it simply check if it is a Variable, could that break somewhere? 

For example tf.keras.layers also keeps their weights as Variables, therefore the taping works. In general, giving a `tf.Variable` as a parameter, I can imagine users to expect this not to be converted to a `Tensor` at all before being applied to `value` (although, of course, this does not have a huge, or not at all, effect for many usecases). Still, quite unexpected behavior to be honest.

(Actually, I went so far to convert any parameter (for my pdfs) _to_ a tf.Variable. In case of a simple value, it's a fixed tf.Variable. Just for consistency)

**Comment by davmre** [comment]
> As I build pdfs myself, not using tf.identity, I am quite interested to know why this is there? What's the purpose of that?

To give a more direct answer: the identity ops are there to support the runtime checks performed by `validate_args`. In order for the checks to actually be run, there needs to be a node in the graph whose value formally depends on them, and the idiomatic way to do this is to add a `tf.identity` op within a `control_dependencies` context.

I think that what your use case requires in general is a guarantee that no ops whatsoever are created in Distribution constructors -- to create a Distribution outside of a GradientTape and use it inside, you need all of the actual op creation to happen at the time when Distribution methods are called. This isn't a guarantee that the Distributions API currently provides. Doing that is an aspiration for the future, though not a very concrete one yet AFAIK -- there are Distributions where that's a more complex to change than it is here.

As a workaround, you can always call `copy()` to reinstantiate a distribution from its original arguments ; this re-calls the constructor so that any ops it creates get recorded in the current context:
```python
var1, var2 = tf.Variable(1.), tf.Variable(2.)
base_normal = tfd.Normal(loc=var1, scale=var2)
with tf.GradientTape() as tape:
    normal = base_normal.copy()
    prob = normal.prob(42.)
print(tape2.watched_variables())  # should contain [var1, var2]
```

**Comment by mayou36** [comment]
Sorry for the late reply. Yes, I see the points, they make sense.

> As a workaround, you can always call `copy()` to reinstantiate a distribution from its original arguments. This re-calls the constructor so that any ops it creates get recorded in the current context:

We currently don't have access to the distribution, just to the method. And even with access it's not "possible" (we don't use the distributions directly but somewhere packed up and mixed with other pdfs. They are a merely implementation detail).

Given that design choice, do you have any plans for (TF 2.0 and) lazy evaluation? I think GradientTape will be a more important part in this scenario and a copy may often is not possible? Or just let it come?

**Comment by davmre** [comment]
If you're just passing around methods, maybe you could pass around `log_prob = lambda x : Normal(loc=var1, scale=var2).log_prob(x)` or similar? (or generically `lambda x : distribution.copy().log_prob(x)`)

We're still thinking through the best ways to adapt TFP APIs including Distributions to work more idiomatically in TF 2. It's not totally trivial since we have a lot of existing code, and maintaining performant graph-mode semantics (via `@tf.function`) is still important in TF 2. I don't believe there's a concrete plan yet, but it's certainly an important concern, and suggestions and thoughts are (as always) appreciated.

**Comment by mayou36** [comment]
I could give that a try. Not very elegant, but if it works... I'll try.

Agree, the graphs will stay quite important. Sure, I'll let you know in case I have an idea, seems non-trivial.

Thanks for the help anyway!
